### PR TITLE
feat(qwen3-asr): add training tracking and fix multi-GPU fine-tuning

### PR DESCRIPTION
**Pull request by lpan-b10**
Qwen3-ASR fine-tuning can fail on its first multi-GPU backward pass when reentrant gradient checkpointing revisits the audio encoder. Use non-reentrant checkpointing, right-pad training batches so labels cover the reference tokens, and let Trainer normalize the model's micro-batch mean loss during gradient accumulation.

The previous collator assumed the prompt started at position zero, but Qwen defaults to left padding. In a mixed-length batch, masking the first `prefix_len` positions could mask padding and leave prompt tokens in the loss. Right padding aligns the prompt with that mask. Earlier losses affected by this bug are not directly comparable with corrected losses. The initial tracking smoke used one sample per GPU, so it did not exercise this padding case.

Add optional W&B and TensorBoard reporting through `REPORT_TO=none|wandb|tensorboard`, disabled by default. Install the selected tracker in the training container and inject the W&B API key through a Baseten secret. Document onboarding, `baseten train` commands, TensorBoard artifact download/viewing, and the distinction between training loss and ASR quality metrics.

The single-GPU preset requests 8 CPUs / 64 GiB RAM, and the launcher checks the allocated GPU count before setup. A live request with the prior 16 CPU / 96 GiB settings selected 2 H100s despite requesting one; the guard caught this before training. Also remove the unused `hf-transfer` dependency and leave `HF_HUB_ENABLE_HF_TRANSFER` unset. The README distinguishes the default `load_dataset` preparation from custom `snapshot_download` loaders; no download speedup is claimed for the default recipe.

Validation after review fixes on exact head `cc4cd67bfe9902595bedce069ed40baeef4f07b5`, in fde-internal / team lei:

| Tracker | GPUs | Training job | Final train / dev loss | Result |
| --- | --- | --- | --- | --- |
| tensorboard | 1 H100 | [31mvgg3](https://app.baseten.co/training/wd187gw/logs/31mvgg3) | 0.2364 / 0.1432 | Passed; checkpoint sync completed |
| tensorboard | 2 H100 | [wd62pk3](https://app.baseten.co/training/wd187gw/logs/wd62pk3) | 0.2342 / 0.1424 | Passed; checkpoint sync completed |
| wandb | 2 H100 | [q49jg9w](https://app.baseten.co/training/wd187gw/logs/q49jg9w) | 0.2342 / 0.1424 | Passed; checkpoint sync completed |

All three reruns used 32 training / 8 development clips from LibriSpeech train.100, four optimizer steps, per-device batch 2, effective batch 8, and gradient checkpointing enabled. The config, launcher, trainer, preparation script, and requirement files matched the PR byte-for-byte, verified by SHA256 in each job. Separate smoke wrappers only bound the run and verify outputs.

Verified finite loss, expected GPU allocation, exact reference-plus-EOS label masking with the real processor, final checkpoint reload and English development transcription, and synchronized final weights plus verification artifacts. Both TensorBoard runs passed in-container and downloaded-event readback with four training and two development points. The W&B run passed authenticated remote history readback with the same point counts and finished state; its run name containing spaces was preserved.

[W&B run](https://wandb.ai/baseten-training/qwen3-asr-medical-transcription/runs/pr138cc4cd67wb2). [Shared training CI passed on this same head](https://github.com/basetenlabs/ml-cookbook/actions/runs/35778603462). Seven local configuration/launcher tests, Ruff, shell syntax, and diff checks also passed. The model-specific collator unit test and dedicated CI job are removed; the shared CI workflow is unchanged by this PR.

These reruns verify the generic cookbook route and tracking, not WER improvement or the custom frozen-medical loader. Earlier medical smoke evidence remains available: [2-H100 TensorBoard](https://app.baseten.co/training/qklg783/logs/32oky0w) and [2-H100 W&B](https://app.baseten.co/training/qklg783/logs/w6vkr0q). Reporting-disabled mode was verified earlier in [qr80yrw](https://app.baseten.co/training/wd187gw/logs/qr80yrw). Full medical training remains pending.

Related: [FDE-3281](https://linear.app/baseten/issue/FDE-3281), [FDE-3179](https://linear.app/baseten/issue/FDE-3179).

### CODE REVIEW

File: examples/qwen3-asr-transformers/training/qwen3_asr_sft.py [hunk 1]
```diff
@@ -129,13 +129,15 @@ def __call__(self, features: List[Dict[str, Any]]) -> Dict[str, torch.Tensor]:
             audio=audios,
             return_tensors="pt",
             padding=True,
+            padding_side="right",
```
[the remark this answers is not in the document]
Comment (lpan-b10) [reply examples/qwen3-asr-transformers/training/qwen3_asr_sft.py:132]: Agreed that this is a training-correctness fix. The PR description and README explain the masking error and warn that affected pre-fix losses are not directly comparable. Our initial tracking smoke used one sample per GPU, which explains the coverage gap. The corrected medical and single-GPU smoke tests verified exact reference transcript plus EOS supervision with the real processor on mixed-length batches. Per the requested scope, cc4cd67 removes the dedicated Qwen unit-test CI job and collator unit test; the padding fix remains, with label validation in the bounded smoke procedure and the existing shared training CI retained.

File: examples/qwen3-asr-transformers/training/config.py [hunk 2]
```diff
@@ -51,8 +75,8 @@
 
 training_compute = definitions.Compute(
     node_count=1,
-    cpu_count=16,
-    memory="96Gi",
+    cpu_count=8 if GPU_COUNT == 1 else 16,
+    memory="64Gi" if GPU_COUNT == 1 else "96Gi",
```
[the remark this answers is not in the document]
Comment (lpan-b10) [reply examples/qwen3-asr-transformers/training/config.py:79]: Fixed in f6f10bb. The default compute table now lists 8 CPU / 64 GiB for one GPU and explicitly notes 16 CPU / 96 GiB for the multi-GPU presets. Removed the inaccurate RAM-threshold justification from the table.

File: examples/qwen3-asr-transformers/README.md [hunk 3]
```diff
@@ -206,6 +346,45 @@ it retains gradients while processing each micro-batch sequentially. Ordinary
...
+run and `SAVE_STRATEGY=steps SAVE_STEPS=2` to exercise evaluation and checkpoint
+writing. Confirm finite losses, completed optimizer steps, TensorBoard events
+when enabled, and a successful reload of `final/`. A tracking test with
+checkpointing disabled does not validate the checkpointed DDP path.
+
+### Dataset download progress
+
+The launcher unsets `HF_HUB_ENABLE_HF_TRANSFER` so Hub snapshot downloads can
+parallelize across files. In `huggingface-hub` 0.36.x, enabling that flag selects
+a serial file loop even when `snapshot_download(max_workers=16)` is requested.
```
[the remark this answers is not in the document]
Comment (lpan-b10) [reply examples/qwen3-asr-transformers/README.md]: Correct: the measured acquisition path was a separate frozen-medical-data loader using snapshot_download(max_workers=16), not this recipe's prepare.py/load_dataset path. The medical subset smoke acquired 544 selected audio files in about 36 seconds on its first run; that is not a controlled speedup comparison with the earlier full-dataset download. In f6f10bb, the README now distinguishes the two paths and explicitly states that no download speedup has been established for the default recipe. Also removed the unused direct hf-transfer dependency.

File: examples/qwen3-asr-transformers/training/run.sh [hunk 4]
```diff
@@ -8,6 +25,9 @@ fi
 
 source .venv/bin/activate
 python -m pip install -q -r requirements.txt
+if [ "${REPORT_TO}" != none ]; then
+    python -m pip install -q -r "requirements.${REPORT_TO}.txt"
+fi
```
[the remark this answers is not in the document]
Comment (lpan-b10) [reply examples/qwen3-asr-transformers/training/run.sh:30]: Changed in f6f10bb to one pip invocation containing the base requirements and, when enabled, the selected tracker requirements. The launcher test now asserts exactly one installation call and its complete argument list for none, wandb, and tensorboard. One clarification: qwen-asr==0.0.6 already pins transformers==4.57.6 and accelerate==1.12.0 exactly in its package metadata, so those versions are constrained transitively.

File: examples/qwen3-asr-transformers/README.md [hunk 5]
```diff
@@ -85,6 +91,140 @@ batch size of 8, every preset therefore keeps the global effective batch at
...
+baseten train checkpoint files --job-id YOUR_JOB_ID --output json > checkpoint-urls.json
+python3 - <<'PY'
+import json
+from pathlib import Path
+from urllib.request import urlretrieve
+
+output = Path("downloaded-events")
+output.mkdir(exist_ok=True)
+events = [item for item in json.loads(Path("checkpoint-urls.json").read_text())
+          if Path(item["relative_file_name"]).name.startswith("events.out.tfevents.")]
```
[the remark this answers is not in the document]
Comment (lpan-b10) [reply examples/qwen3-asr-transformers/README.md:180]: Verified against the real Baseten CLI v1.0.0: `baseten train checkpoint files --job-id qe6zo13 --output json` returns a bare JSON array, not the REST response wrapper. The latest readback returned 37 entries, including the TensorBoard event file; the README's download example has also been executed successfully. Keeping the current parsing because it matches the CLI output. This readback verifies the output shape and event download; it does not claim a multi-page stress test.

File: examples/qwen3-asr-transformers/training/test_tracking.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        for mode in ("none", "wandb", "tensorboard"):
+            with self.subTest(mode=mode), tempfile.TemporaryDirectory() as directory:
+                root = Path(directory)
+                venv = root / ".venv"
+                (venv / "bin").mkdir(parents=True)
+                (venv / "bin/activate").write_text(f'export PATH="{venv}/bin:$PATH"\n')
+                fake_python = venv / "bin/python"
+                fake_python.write_text(
+                    '#!/bin/bash\nprintf "%s\\n" "$*" >> calls.txt\n'
+                )
```
[the remark this answers is not in the document]
Comment (lpan-b10) [reply examples/qwen3-asr-transformers/training/test_tracking.py:110]: Fixed in f6f10bb. The stub now writes sys.argv[1:] as JSON arrays, and the test checks the argument immediately following --run_name equals the complete `run with spaces` value. I verified a temporary mutation removing the shell quoting fails this test for both enabled trackers. The test also checks the exact combined pip arguments.

File: examples/qwen3-asr-transformers/training/run.sh [hunk 7]
```diff
@@ -1,5 +1,22 @@
...
+
+GPU_COUNT="${BT_NUM_GPUS:-${NUM_GPUS:-1}}"
+if ! [[ "${GPU_COUNT}" =~ ^[1-9][0-9]*$ ]]; then
+    echo "GPU count must be a positive integer; got: ${GPU_COUNT}" >&2
+    exit 1
+fi
+if [ -n "${EXPECTED_GPU_COUNT:-}" ] && [ "${GPU_COUNT}" != "${EXPECTED_GPU_COUNT}" ]; then
+    echo "Expected ${EXPECTED_GPU_COUNT} GPUs, allocated ${GPU_COUNT}; check the compute request" >&2
+    exit 1
+fi
```
[the remark this answers is not in the document]
Comment (lpan-b10) [reply examples/qwen3-asr-transformers/training/run.sh:19]: Confirmed: BT_NUM_GPUS reflects the selected instance's GPU count. The training worker template injects it from resources.instance.gpu_count. We also exercised this live: job 3yp72n3 requested one H100 with 16 CPU / 96 GiB, received an H100:2 instance, and the guard stopped before dependency installation or training. The adjusted 8 CPU / 64 GiB request in job qe6zo13 received exactly one H100 and completed training, reload/transcription, and checkpoint synchronization. Retaining the guard based on both source and observed behavior.

File: examples/qwen3-asr-transformers/README.md [hunk 8]
```diff
@@ -85,6 +91,140 @@ batch size of 8, every preset therefore keeps the global effective batch at
...
+
+The URLs expire; regenerate them if needed. Keep `checkpoint-urls.json` private
+and delete it after downloading.
+
+To view downloaded logs on your laptop, install TensorBoard in a separate local
+virtual environment:
+
+```bash
+python3 -m venv .tensorboard-venv
+.tensorboard-venv/bin/python -m pip install tensorboard==2.20.0 setuptools==80.9.0
```
[the remark this answers is not in the document]
Comment (lpan-b10) [reply examples/qwen3-asr-transformers/README.md:196]: Keeping this compatibility pin. TensorBoard 2.20.0 still imports pkg_resources, which setuptools removed in v82; `setuptools>=41.0.0` permits versions that no longer provide it. Version 80.9.0 was verified with the local viewer. Added this rationale to the README in f6f10bb. Source: https://setuptools.pypa.io/en/latest/deprecated/pkg_resources.html

File: examples/qwen3-asr-transformers/training/qwen3_asr_sft.py [hunk 9]
```diff
@@ -316,7 +328,9 @@ def main():
         fp16=not use_bf16,
         ddp_find_unused_parameters=False,
         remove_unused_columns=False,
-        report_to="none",
+        report_to=args_cli.report_to,
+        run_name=args_cli.run_name,
+        logging_dir=os.path.join(args_cli.output_dir, "tensorboard"),
```
[the remark this answers is not in the document]
Comment (lpan-b10) [reply examples/qwen3-asr-transformers/training/qwen3_asr_sft.py:333]: Yes. For completed job qe6zo13, checkpoint synchronization finished and the storage listing included `tensorboard/events.out.tfevents...`. We downloaded that event file and independently read it locally with EventAccumulator: four train/loss points and two eval/loss points, ending at 0.2364 and 0.143238 respectively. Earlier job wlvnyvq also passed the README's exact download example and local TensorBoard HTTP readback. These results are recorded in the PR validation table; no directory-sync change is needed.

**Review by A-K-Erol — APPROVED** [review]
